mllib/src/main/scala/org/apache/spark/mllib/regression/IsotonicRegression.scala

```diff
@@ -0,0 +1,238 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.mllib.regression
+
+import java.io.Serializable
+import java.util.Arrays.binarySearch
+
+import org.apache.spark.api.java.{JavaDoubleRDD, JavaRDD}
+import org.apache.spark.rdd.RDD
+
+/**
+ * Regression model for Isotonic regression
+ *
+ * @param features Array of features.
+ * @param labels Array of labels associated to the features at the same index.
+ */
+class IsotonicRegressionModel (
+    features: Array[Double],
+    val labels: Array[Double])
+  extends Serializable {
+
+  /**
+   * Predict labels for provided features
+   * Using a piecewise constant function
+   *
+   * @param testData features to be labeled
+   * @return predicted labels
+   */
+  def predict(testData: RDD[Double]): RDD[Double] =
+    testData.map(predict)
+
+  /**
+   * Predict labels for provided features
+   * Using a piecewise constant function
+   *
+   * @param testData features to be labeled
+   * @return predicted labels
+   */
+  def predict(testData: JavaRDD[java.lang.Double]): JavaDoubleRDD =
+    JavaDoubleRDD.fromRDD(predict(testData.rdd.asInstanceOf[RDD[Double]]))
+
+  /**
+   * Predict a single label
+   * Using a piecewise constant function
+   *
+   * @param testData feature to be labeled
+   * @return predicted label
+   */
+  def predict(testData: Double): Double = {
+    val result = binarySearch(features, testData)
+
+    val index =
+      if (result == -1) {
+        0
+      } else if (result < 0) {
+        -result - 2
+      } else {
+        result
+      }
+
+    labels(index)
+  }
+}
+
+/**
+ * Isotonic regression
+ * Currently implemented using parallel pool adjacent violators algorithm
+ */
+class IsotonicRegression
+  extends Serializable {
+
+  /**
+   * Run algorithm to obtain isotonic regression model
+   *
+   * @param input (label, feature, weight)
+   * @param isotonic isotonic (increasing) or antitonic (decreasing) sequence
+   * @return isotonic regression model
+   */
+  def run(
+      input: RDD[(Double, Double, Double)],
+      isotonic: Boolean = true): IsotonicRegressionModel = {
```

Comment:
The default argument value is not Java compatible and we don't use this kind of API in `spark.mllib`. The class `IsotonicRegression` should have a parameter called `isotonic`, similar to `k` in `KMeans`. The user code should look like:

~~~
val ir = new IsotonicRegression()
  .setIsotonic(false)
val irModel = ir.run(input)
~~~